### PR TITLE
Moved Nonce Logic to Cart API Classes

### PR DESCRIPTION
**Pull request by mikejolley**
When working on https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3810 I noticed some bugs and discrepancies with some of the cart endpoints.

- Products Route was loading the cart, in order to get a valid nonce response
- Some other endpoints do not load the cart, but still return nonces. They may be invalid.
- Some routes extended the AbstractCartRoute, but do not have Cart Schema. This would break on error, because AbstractCartRoute returns an updated cart on errors.

I've fixed this with a number of small improvements:

1. All routes extending AbstractCartRoute **must** provide cart schema. They can provide an item schema too, but the first param must be cart.
2. Nonce checks and generation are now within the AbstractCartRoute. Only cart and checkout requests need to return a nonce, or validate it.

### How to test the changes in this Pull Request:

1. PHP Unit tests will pass.
2. Add to cart in incognito and checkout successfully

### Changelog

> Improved nonce logic by moving nonces to cart routes only.

### CODE REVIEW

File: src/StoreApi/Routes/AbstractCartRoute.php [hunk 1]
```diff
@@ -2,13 +2,41 @@
...
+	protected $cart_schema;
+
+	/**
+	 * Constructor accepts two types of schema; for for the item being returned, and one for the cart as a whole. These
+	 * may be the same depending on the route.
+	 *
+	 * @param CartSchema     $cart_schema Schema class for the cart.
+	 * @param AbstractSchema $item_schema Schema class for this route's items if it differs from the cart schema.
+	 */
+	public function __construct( CartSchema $cart_schema, AbstractSchema $item_schema = null ) {
```
Comment (mikejolley) [inline src/StoreApi/Routes/AbstractCartRoute.php:35]: This ensures we have a cart schema regardless of the route.

File: src/StoreApi/Routes/AbstractCartRoute.php [hunk 2]
```diff
@@ -18,7 +46,37 @@ abstract class AbstractCartRoute extends AbstractRoute {
...
+			$response = $this->get_route_error_response( 'unknown_server_error', $error->getMessage(), 500 );
+		}
+
+		if ( is_wp_error( $response ) ) {
+			$response = $this->error_to_response( $response );
+		}
+
+		$response->header( 'X-WC-Store-API-Nonce', wp_create_nonce( 'wc_store_api' ) );
+		$response->header( 'X-WC-Store-API-Nonce-Timestamp', time() );
+		$response->header( 'X-WC-Store-API-User', get_current_user_id() );
```
Comment (mikejolley) [inline src/StoreApi/Routes/AbstractCartRoute.php:67]: Only cart routes need this.

File: src/StoreApi/Routes/AbstractCartRoute.php [hunk 3]
```diff
@@ -69,11 +155,22 @@ protected function get_route_error_response( $error_code, $error_message, $http_
 						$additional_data,
 						[
 							'status' => $http_status_code,
-							'cart'   => $this->schema->get_item_response( $cart ),
+							'cart'   => $this->cart_schema->get_item_response( $cart ),
```
Comment (mikejolley) [inline src/StoreApi/Routes/AbstractCartRoute.php:158]: This was a bug - some routes did not have cart schema.

File: src/StoreApi/Routes/CartUpdateCustomer.php [hunk 4]
```diff
@@ -75,17 +48,17 @@ public function get_args() {
...
+						'validate_callback' => [ $this->schema->billing_address_schema, 'validate_callback' ],
 					],
 					'shipping_address' => [
 						'description'       => __( 'Shipping address.', 'woo-gutenberg-products-block' ),
 						'type'              => 'object',
 						'context'           => [ 'view', 'edit' ],
-						'properties'        => $this->shipping_address_schema->get_properties(),
-						'sanitize_callback' => [ $this->shipping_address_schema, 'sanitize_callback' ],
-						'validate_callback' => [ $this->shipping_address_schema, 'validate_callback' ],
+						'properties'        => $this->schema->shipping_address_schema->get_properties(),
```
Comment (mikejolley) [inline src/StoreApi/Routes/CartUpdateCustomer.php:59]: These are now public and accessible through cart.

File: src/StoreApi/Routes/Checkout.php [hunk 5]
```diff
@@ -39,23 +39,13 @@ public function get_path() {
...
-			$this->check_nonce( $request );
-			$response = parent::get_response( $request );
-		} catch ( RouteException $error ) {
-			$response = $this->get_route_error_response( $error->getErrorCode(), $error->getMessage(), $error->getCode() );
-		} catch ( Exception $error ) {
-			$response = $this->get_route_error_response( 'unknown_server_error', $error->getMessage(), 500 );
-		}
-		return $response;
+	protected function requires_nonce( \WP_REST_Request $request ) {
+		return true;
```
Comment (mikejolley) [inline src/StoreApi/Routes/Checkout.php:48]: All checkout endpoints require a nonce.

File: src/StoreApi/Routes/AbstractCartRoute.php [hunk 6]
```diff
@@ -2,13 +2,41 @@
...
+
+	/**
+	 * Schema class for the cart.
+	 *
+	 * @var CartSchema
+	 */
+	protected $cart_schema;
+
+	/**
+	 * Constructor accepts two types of schema; for for the item being returned, and one for the cart as a whole. These
```
Comment (budzanowski) [inline src/StoreApi/Routes/AbstractCartRoute.php]: for for

**Review by budzanowski — APPROVED** [review]
Looks good.

**Review by senadir — APPROVED** [review]
Code looks legit, tests fine.

**Pull request merged**